src/common/emitter.ts

```diff
@@ -32,8 +38,21 @@ export class Emitter<T> {
   /**
    * Emit an event with a value.
    */
-  public emit(value: T): void {
-    this.listeners.forEach((cb) => cb(value))
+  public async emit(value: T): Promise<void> {
+    let resolve: () => void
+    const promise = new Promise<void>((r) => (resolve = r))
```

Comment:
I think we have to separate it because if we call `cb` in the promise callback then `promise` will still be undefined as it hasn't been set yet (since these callbacks are ran synchronously).

Reply:
Like if we did something like:
```
const promise = Promise.all(listeners.map(async (cb) => {
  await cb(value, promise)
})
```

Reply:
Ohhh, I gotchu.!